File: scripts/python/parsers/dashboard_parser.py
"""Parser for Dashboard metadata (.dashboard-meta.xml), nested in folder hierarchy."""
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_dashboards(type_dir: str) -> dict:
    """
    @param type_dir: path to the dashboards/ folder (dashboards are nested in sub-folders)
    @returns: { dashboards: list }
    """
    dashboards = []
    p = Path(type_dir)
    if not p.exists():
        return {'dashboards': dashboards}

    for f in sorted(p.rglob('*.dashboard-meta.xml')):
        xml = f.read_text(encoding='utf-8')
        dashboards.append({
            'name':        f.name.replace('.dashboard-meta.xml', ''),
            'title':       _extract_tag(xml, 'title'),
            'description': _extract_tag(xml, 'description'),
        })

    return {'dashboards': dashboards}

File: scripts/node/parsers/dashboardParser.js
'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

/**
 * Reads Dashboard metadata from SFDX source format (dashboards are nested in folders).
 * @param {string} typeDir - path to the dashboards/ folder
 * @returns {{ dashboards: Array }}
 */
async function parseDashboards(typeDir) {
  const dashboards = [];
  if (!fs.existsSync(typeDir)) return { dashboards };

  const walk = (dir) => {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(full);
      } else if (entry.name.endsWith('.dashboard-meta.xml')) {
        const xml = fs.readFileSync(full, 'utf8');
        dashboards.push({
          name:        entry.name.replace('.dashboard-meta.xml', ''),
          title:       extractTag(xml, 'title'),
          description: extractTag(xml, 'description'),
        });
      }
    }
  };

  walk(typeDir);
  return { dashboards };
}

module.exports = { parseDashboards };
